src/features/bivariate_color_manager/atoms/bivariateColorManager.ts

```diff
@@ -16,25 +32,128 @@ export type LayerSelectionInput = {
 
 export type BivariateColorManagerAtom = typeof bivariateColorManagerAtom;
 
+const DEFAULT_STATE = {
+  _initialData: null,
+  filteredData: null,
+  indicators: null,
+  filters: { layers: null },
+  selectedRows: {},
+  layersSelection: null,
+};
+
 export const bivariateColorManagerAtom = createAtom(
   {
     setLayersSelection: (input: LayerSelectionInput) => input,
+    runFilters: () => undefined,
+    setLayersFilter: (input: SelectItemType | null) => input,
+    setSelectedRows: (key: string) => key,
     bivariateColorManagerResourceAtom,
   },
   (
-    { onAction, onChange },
-    state: BivariateColorManagerAtomState = {
-      data: null,
-    },
+    { onAction, onChange, schedule, create },
+    state: BivariateColorManagerAtomState = DEFAULT_STATE,
   ) => {
     onChange('bivariateColorManagerResourceAtom', (resource) => {
-      if (!resource.data) return (state = { data: null });
+      if (!resource.data) return (state = DEFAULT_STATE);
+
+      const { bivariateColorManagerData, indicators } = resource.data;
 
       state = {
-        data: resource.data,
+        ...DEFAULT_STATE,
+        _initialData: bivariateColorManagerData,
+        filteredData: bivariateColorManagerData,
+        indicators: indicators,
       };
 
-      if (!state.data) state.data = null;
+      if (!state._initialData) {
+        state = DEFAULT_STATE;
+      }
+    });
+
+    onAction('setLayersFilter', (input) => {
+      state.filters = { ...state.filters, layers: input?.value || null };
+      schedule((dispatch) => {
+        dispatch(create('runFilters'));
+      });
+    });
+
+    onAction('setSelectedRows', (key) => {
+      state = {
+        ...state,
+        selectedRows: {
+          ...state.selectedRows,
+          [key]: !state.selectedRows[key],
+        },
+      };
+    });
+
+    onAction('runFilters', () => {
+      const filterFunctionsToApply = Object.entries(state.filters)
+        .map(([key, value]) => (value ? [filterFunctions[key], value] : null))
+        .filter(Boolean) as [FilterFunction, FiltersValues][];
+
+      if (filterFunctionsToApply.length === 0 || !state._initialData) {
+        state = {
+          ...state,
+          filteredData: state._initialData,
+
+          // drop selected rows/sublist after filter cleared
+          selectedRows: {},
+          layersSelection: null,
+        };
+        return;
+      }
+
+      const initialData = state._initialData;
+      const nextFilteredData = Object.keys(initialData).reduce((acc, key) => {
+        const value = { ...initialData[key] };
+
+        // check if row passes all filter functions
+        const filterPassed = filterFunctionsToApply.every(
+          ([filterFunction, filterValue]) =>
+            filterFunction(key, value, filterValue),
+        );
+
+        if (filterPassed) {
+          // do postprocessing for passed row here
+
+          // if filter layer selected - stay only filtered layer in row's sublist
```

Comment:
A little bit confused that specific layer filtering logic take about a half of generic applyFilters method while there is already dedicated function **layersFilterFunction** for such purposes

Reply:
@andrei-kucherov the reason I've splitted it into 2 parts is next: 
- filterFunctionsToApply as an array of generic filter functions that will we applied to data array to filter it (it's pure)
- but we also need to modify it, it's not just filtering, so I decided to separate it to post filtering phase (not pure)
just don't want to mix pure and not pure staff, as it can lead to bugs

so as I see this: 
you need to add new filter - just add filter logic to filterFunctions and add filter body to the atom's store
if you need to modify smth after filtering - add postprocessing logic when specific filter applied

also as a case it's possible to implement this filtering sublist on component level
so we can get rid of this postprocessing part here 
but not sure it's a best approach